geoh5py/shared/entity_type.py

```diff
@@ -20,136 +20,209 @@
+        self.uid = self._modify_attribute("ID", uid, **kwargs)
+        self.description = self._modify_attribute("Description", description, **kwargs)
+        self.name = self._modify_attribute("Name", name, **kwargs)
+        self._on_file = False
 
         self.workspace.register(self)
 
+    @classmethod
+    def _modify_attribute(cls, attribute_name: str, attribute: Any, **kwargs):
+        """
+        Modify the attribute based on the name and kwargs.
+
+        :param attribute_name: The name of the attribute to modify.
+        :param attribute: The attribute to modify.
+        :param kwargs: The kwargs to modify the attribute.
+
+        :return: The modified attribute.
+        """
+        if cls._attribute_map.get(attribute_name, None):
+            if attribute_name in kwargs:
+                return kwargs[attribute_name]
+        return attribute
+
+    def _set_workspace(self, workspace: Workspace) -> Workspace:
+        """
+        Set the workspace for the entity type.
+        It is private as workspace should not be changed
+
+        :param workspace: The workspace to set.
+        """
+        if not hasattr(workspace, "create_entity"):
+            raise TypeError(f"Workspace must be a Workspace, not {type(workspace)}")
+
+        workspace_ = weakref.ref(workspace)()
+
+        if workspace_ is None:
+            raise ValueError("Workspace is not available.")
+
+        return workspace_
+
     @property
-    def attribute_map(self):
+    def attribute_map(self) -> dict[str, str]:
         """
-        :obj:`dict` Correspondence map between property names used in geoh5py and
+        Correspondence map between property names used in geoh5py and
         geoh5.
         """
         return self._attribute_map
 
+    def copy(self, **kwargs) -> EntityType:
+        """
+        Copy this entity type to another workspace.
+        """
+
+        attributes = {
+            prop: getattr(self, prop)
+            for prop in dir(self)
+            if isinstance(getattr(self.__class__, prop, None), property)
+            and getattr(self, prop) is not None
+        }
+
+        attributes.update(kwargs)
+
+        if attributes.get("uid") in getattr(
+            attributes.get("workspace", self.workspace), "_types"
+        ):
+            del attributes["uid"]
+
+        return self.__class__(**attributes)
+
     @property
     def description(self) -> str | None:
+        """
+        The description of the entity type.
+        """
         return self._description
 
     @description.setter
-    def description(self, description: str):
+    def description(self, description: str | None):
+        if not isinstance(description, (str, type(None))):
+            raise TypeError(
+                f"Description must be a string or None, find {type(description)}"
+            )
+
         self._description = description
-        self.workspace.update_attribute(self, "attributes")
+
+        if hasattr(self, "_on_file"):
```

Comment:
```suggestion
        if self._on_file:
```

Reply:
This method quite hacky. 
maybe us8ing pydantic will solve the issue

Reply:
set register in workspace setter